### ISSUE TITLE
False negative 'Down' for VPS running a dozen monitored services behind an NGINX reverse proxy

### ISSUE THREAD

**Issue by lightweight** [louislam/uptime-kuma#3383]
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### 📝 Describe your problem

I have an up-to-date Uptime Kuma instance (deployed via Docker on a colocated server running Ubuntu Linux 22.04) which is monitoring web services on a number of different VPSs.  One of those VPSs has a dozen services being monitored (the most of any of the VPSs). The services are all deployed via Docker Compose and are running behind an NGINX instance which provides reverse proxying. Most of the time, all 12 services show as 'Down', each with the message `[<service name>] [DOWN] connect ECONNREFUSED <IPaddress>`. 

I should note that all these services are entirely usable and 'up' except, apparently from the perspective of the server hosting the Uptime Kuma instance (a `wget` of any site that's shown as 'Down' by Uptime Kuma also gives a 'failed: Connection refused' error if run from the server's commandline).  

Then, for an occasional period of half an hour (almost exactly), they all show as 'Up'. After which, they are again 'Down' for a few hours. 

It seems like some sort of rate-limiting mechanism on the server being monitored is being tripped by the dozen service checks every minute, which blocks access to them for a period of time. When that rate-limiting period ends, they're visible again... but I cannot, for the life of me, work out *what mechanism* is providing this rate-limiting! It doesn't seem to be NGINX (which is, of course, my first guess) or Fail2Ban or something like that... 

I'm running two instance of Uptime Kuma this way, and one has never had these sorts of false negatives (although it's not monitoring so many services on any one server) and the other has no false natives from any other hosts. 
Has anyone else experienced this sort of false negative and have any insights?

### 🐻 Uptime-Kuma Version

1.22.1 (louislam/uptime-kuma:1)

### 💻 Operating System and Arch

Ubuntu 22.04 host system, default docker containers

### 🌐 Browser

all

### 🐋 Docker Version

Docker 24.0.1 and Docker Compose v2.18.1

**Comment by rezzorix** [comment]
Sounds less like uptime-kuma problem and more like lack of configuration of firewall or other security settings.

Which Provider is your VPS on?
Could be that their security / firewall system sees the services being monitored as some kind of threat and then blocks the monitoring VPS ip…

Have you tried whitelisting the IP monitoring the services?

**Comment by lightweight** [comment]
Thanks for the response @rezzorix - my VPS is on Hetzner - I've already asked them if there's anything in their infrastructure that be rate limiting my access from one server to another in their network. They assure me that there is not... I've checked my firewall, ufw (no evidence of blocking anything?) and my NGINX instance (it's capable of doing rate limiting, but also haven't been able to see evidence of it doing any - it's not configured to do any, unless there's a default value, but I haven't been able to find it)... There's no where I can think of where whitelisting would be relevant... :(   My next step is to see if I can explicitly tell NGINX *not* to rate limit (in case there's a default rate limit trigger in place)...

**Comment by rezzorix** [comment]
In addition to the software firewall of the operation system (in your case UFW) there should be also another option for a lets call it “hardware” firewall in your Hetzer interface. 

In the Hetzner documentation they state that all incoming ports are closed by default if you do not set any specific rules. 

Therefore you would need to open ports in this “hardware” firewall, too. 
As I dont use Hetzner I can only guess, but usually this is done somewhere under the security or networking options.

**Comment by rezzorix** [comment]
I just wrote this and forgot that you have already stated that services are up sometimes… In that case it must be rate limiting the monitoring ip somewhere. Since you are using Nginx and it is across the board of all your services, that would be the first to fix. How about trying another reverse proxy just to see if it is because of Nginx?

**Comment by lightweight** [comment]
Hah! I think I found it: it looks like the UFW firewall on my VPS has default rate limiting... like in `/etc/ufw/before.rules`:

`-A ufw-not-local -m limit --limit 3/min --limit-burst 10 -j ufw-logging-deny` 

I suspect that the `--limit-burst 10` argument (it also appears in `user.rules` and `user6.rules`) is the thing that makes my server with 12 service fail, but the ones with < 10 are all fine. I've changed 10 to 20 and am waiting for the limit to expire... will confirm if it's fixed the issue when it does...

**Comment by lightweight** [comment]
Ok, that didn't work as hoped... The other option I've tried is 'whitelisting' the Uptime Kuma IP with the host doing the blocking... awaiting confirmation it's worked. Seems this is a problem for some other similar (but not libre) services, e.g. https://betterstack.com/community/questions/whitelisting-better-uptime-ip-ufw/

**Comment by vinc3nz** [comment]
Hey @lightweight,
Were you able to fix the problem? If so, please close the issue.

**Comment by lightweight** [comment]
Sadly, despite thinking I've figured out what's causing the problem - the UFW firewall - I haven't yet been able to fix it.

**Comment by lightweight** [comment]
Boom! Finally, I've figured it out! On the server that's chronically non-responsive, issuing the ECONNREFUSED notifications for extended periods of time (resulting in false notifications of failed services), but occasionally working briefly... 

Turns out, I'm also running an instance of Mailcow (dockerized). It has a 'Fail2Ban' implementation as one of its many services, designed to block IPs engaging in nefarious behaviours. Turns out that a) my Uptime Kuma instance was trigging the Fail2Ban block, and b) *somehow* that blockage was not limited to the Mailcow service, but was somehow affecting all the other services on the same VPS that were using the Nginx reverse proxy. Not sure *how* that was happening. 

On the Fail2Ban interface provided by Mailcow, I could see that both the IPv4 and IPv6 addresses of the Uptime Kuma server were blocked. I've now whitelisted both and issued an 'unban' for them, and Uptime Kuma is now working as it should! Hazzah! Hope this helps someone!
